### PR TITLE
fix(cmf-cqrs): Handle websocket timeout on creation

### PR DESCRIPTION
**Pull request by sgendre**
**What is the problem this PR is trying to solve?**
On app launch and websocket creation, if the backend service is unavailable, user get notified way too late, something like near 1 minute.
**What is the chosen solution to this problem?**
Align it with the timeout defined for ping-pong heartbeat.

Please note that the mock websocket lib does not provide enough precision to test this use case.

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: packages/cmf-cqrs/src/middleware/smartWebsocket.js [hunk 1]
```diff
@@ -79,6 +80,15 @@ export function startWebsocket(url, offlinebuffer, options) {
...
 	};
+
+	function wsStillNotYetOpened() {
+		if (ws.readyState === WebSocket.CONNECTING) {
+			ws.onpingtimeout();
+		}
+		clearInterval(wsCreationIntervalId);
+	}
+
+	wsCreationIntervalId = setInterval(wsStillNotYetOpened, options.pingTimeoutDelay);
```
Comment (jmfrancois) [inline packages/cmf-cqrs/src/middleware/smartWebsocket.js]: options.pingTimeoutDelay is not required, so please check it exists

File: packages/cmf-cqrs/src/middleware/smartWebsocket.js [hunk 2]
```diff
@@ -79,6 +80,18 @@ export function startWebsocket(url, offlinebuffer, options) {
 		}
 		ws.close();
 	};
+
+	function wsStillNotYetOpened() {
+		if (ws.readyState === WebSocket.CONNECTING) {
+			ws.onpingtimeout();
+		}
+		clearInterval(wsCreationIntervalId);
```
Comment (jsomsanith) [inline packages/cmf-cqrs/src/middleware/smartWebsocket.js]: 1. you init an interval to notify if the ws is still in connecting state
2. **on the first interval call**, you notify if needed, but in all cases, you clear the interval.
I think you forgot an `else`

**Comment by jsomsanith** [comment]
`Please note that the mock websocket lib does not provide enough precision to test this use case.`
--> you init the smart ws, the mock is in CONNECTING state.
--> you use jest fake timer to flush tu timeouts, and check that onpingtimeout() is called

File: packages/cmf-cqrs/src/middleware/smartWebsocket.js (hunk 2 above)
Comment (sgendre) [reply packages/cmf-cqrs/src/middleware/smartWebsocket.js]: If the ws is not in connecting state, the interval becomes useless.
If the ws is in connecting mode, the interval is useless as as ws.onpingtimeout will close teh ws and that another is reopen by design of smartwebsocket.

I will indeed change setInterval to setTimeout

**Comment by sgendre** [comment]
@jsomsanith I tried that but the point is that I can't find a way to make the mock lib staying in CONNECTING state.
Using jest.advanceTimersByTime and praying that mock lib is still before OPEN state seems unreliable to me.

**Comment by jsomsanith** [comment]
Looking quickly at the ws mock code, I only see a ` this.readyState = WebSocket.OPEN;` when you call :
```
send() => delay() => if(server){ ... this.readyState = WebSocket.OPEN; }
```

Did I missed something ?
Did you had a readyState change when you tried to test it with only 
* use `jest.useFakeTimers();`
* instanciate the socket
* call `jest.runAllTimers();`

**Comment by sgendre** [comment]
from what I see `constructor => delay() => if(server){ ... this.readyState = WebSocket.OPEN; or this.readyState = WebSocket.CLOSED; }`

I provide failling tests cases directly depeneding of the ws mock implementation imho, any contribution is welcome.

**Comment by jsomsanith** [comment]
The problem is not in the mock, the constructor doesn't call delay or send.
The problem is that SmartWebSocket send a ping message on open. This calls send() => delay() => set status to OPEN.

The delay of send is set to 4ms. So setting a pingTimeout to 1ms does the trick.

File: packages/cmf-cqrs/src/middleware/smartWebsocket.test.js [hunk 3]
```diff
@@ -187,5 +187,33 @@ describe('smart websocket tests', () => {
...
+				checkInterval: 70,
+				// < 4ms, that is the delay from ws mock to send
+				// smartWebStocket send a ping on connect, so to test the connect timeout
+				// we need to trigger timeout before any send delay
+				pingTimeoutDelay: 1,
+			};
+
+			// when
+			const result = SmartWebsocket(url, options); // eslint-disable-line
+			result.send = () => {};
```
Comment (jmfrancois) [inline packages/cmf-cqrs/src/middleware/smartWebsocket.test.js]: why not jest.fn ? is it called ? should it be ?

File: packages/cmf-cqrs/src/middleware/smartWebsocket.test.js [hunk 4]
```diff
@@ -187,5 +187,33 @@ describe('smart websocket tests', () => {
...
+			expect(result.getReadyState()).toEqual(WebSocket.CONNECTING);
+			jest.runTimersToTime(1);
+
+			// then
+			expect(result.getReadyState()).toEqual(WebSocket.CONNECTING);
+			expect(options.onPingTimeout).toBeCalled();
```
Comment (jmfrancois) [inline packages/cmf-cqrs/src/middleware/smartWebsocket.test.js]: please add this test previous the runTimer to be sure it has not been called

File: packages/cmf-cqrs/src/middleware/smartWebsocket.test.js (hunk 3 above)
Comment (jsomsanith) [reply packages/cmf-cqrs/src/middleware/smartWebsocket.test.js]: Oops that was a test but not needed. I'll remove it

**Comment by sgendre** [comment]
lgtm, and thanks for that 4ms test range !

**Pull request merged**